### PR TITLE
increase page_size in workflows request to API

### PR DESCRIPTION
**Pull request by mcbouslog**
Temporary solution to issue #2819.

API workflow response limited to 20 items, NfN has over 20 active workflows, so new workflows not showing up in Project Builder.

Temporary solution is to increase page_size param from 20 (default) to 50.

### CODE REVIEW

File: app/lib/get-workflows-in-order.coffee [hunk 1]
```diff
@@ -1,5 +1,7 @@
-getWorkflowsInOrder = (project, query) ->
+getWorkflowsInOrder = (project, query = {}) ->
   order = project.configuration?.workflow_order ? []
+  unless query?.page_size?
```
Comment (camallen) [inline app/lib/get-workflows-in-order.coffee]: Not a blocker for me but can't you just set this query param where you need to call it ala [here](https://github.com/zooniverse/Panoptes-Front-End/blob/1f94f898f3480f1659aa47604d3af2196eb26a9b/app/pages/project/home.cjsx#L24)

File: app/pages/project/finished-banner.cjsx [hunk 2]
```diff
@@ -33,7 +34,7 @@ module.exports = React.createClass
     if project.redirect
       Promise.resolve false
     else
-      project.get('workflows').then (allWorkflows) ->
+      getWorkflowsInOrder(project).then (allWorkflows) ->
```
Comment (camallen) [inline app/pages/project/finished-banner.cjsx]: 20 workflows here will be a really slow call....are they really all active and unfinished?

**Comment by mcbouslog** [comment]
@camallen - great point, and I'm totally fine changing accordingly, but it seems like all the workflows requests (except ones requesting a single workflow by including workflow id) would want all workflows? In which case my thinking was instead of adding the `page_size: 50` query to all the requests, add to the component, but I'm admittedly ignorant to any negative consequences this may have to the API, please let me know!

Just for reference, I think the following are all the workflows requests:
1. project home page
2. project classify page (for selecting random workflow)
3. project builder - workflows list (to select workflow to edit)
4. project builder - visibility section (for selecting which workflows are active)
5. stats page
6. admin project status page
7. finished banner component

**Comment by mcbouslog** [comment]
And all workflows presumes there are less than 50.

A long term solution would be to load the first 20, then have a "Load more..." button or something to load additional workflows as needed.

File: app/pages/project/finished-banner.cjsx (hunk 2 above)
Comment (mcbouslog) [inline app/pages/project/finished-banner.cjsx]: yes i should add a `active: true` query at minimum?

**Comment by camallen** [comment]
right - we can modify the default page size for workflows on the api if needed. 

Remember the really slow workflows load times for the NFN landing page? So wherever we are using this we will suffer from the slow req (how many active workflows are they expecting?). If we can we should identify which calls actually need a full workflow representation and just use the fields param (see nfn landing page call) to create fast api responses as they mostly just need ids for some of those routes you've got (maybe more).

**Comment by mcbouslog** [comment]
Ah yes the fields param! Great idea, I'll update accordingly.

**Comment by mcbouslog** [comment]
ok i think the various workflows API calls have been properly updated with the `fields` param, except for the [stats index page](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/pages/project/stats/index.cjsx#L15), where it should be:
```
getWorkflowsInOrder @props.project, {active: true, fields: 'classifications_count,completeness,display_name,retired_set_member_subjects_count,retirement,subjects_count'}
```
but i can't get `subjects_count` back from the API, and i'm not sure why (all the others come back ok). the stats page workflows response is limited by `active: true` though, i think good enough for now? but let me know if you can tell why no luck with `subjects_count`.

i'm excited to see how much this will help with the workflows load-time on NfN! let me know what you guys think. thanks,

**Comment by camallen** [comment]
@mcbouslog can you get this deployed for testing?

**Comment by mcbouslog** [comment]
no prob - [here's a staged branch](https://inc-workflow-response-page-size.pfe-preview.zooniverse.org/).

i have a project on preview i could add you both as collaborators on if you'd like, just let me know your preview usernames.

**Comment by camallen** [comment]
the stats page is failing due to the api not respecting a virtual field on the workfow (`subjects_count`) [here](https://github.com/zooniverse/Panoptes/blob/master/app/controllers/api/v1/workflows_controller.rb#L58), raise issue here https://github.com/zooniverse/Panoptes/issues/1881

**Comment by camallen** [comment]
@mcbouslog i've got a PR up for the virtual attribute via fields param. Once that is merged we should test again. I've played around locally and this is working for me but it needs more eyes / testing as it could break things pretty quick. 

Also i still don't like the page param being added in the get workflows in order component but I agree it could get a bit messy the other way.

**Comment by mcbouslog** [comment]
sounds good @camallen, i'll update the stats page workflows call once the related Panoptes PR is merged.

agreed regarding additional review, i'll see if i can pull one more person in. an error with how workflows are loaded would be very bad.

also agree regarding the `page_size` param, and open to suggestion, but hopefully it's ok for now, until i can implement a paging solution.

**Comment by mcbouslog** [comment]
stats page workflow request updated, and staged branch also updated accordingly.

i've roped @srallen in as well.

let me know what you think. thanks!

**Comment by camallen** [comment]
the stats pages are working just fine now, https://inc-workflow-response-page-size.pfe-preview.zooniverse.org/projects/zooniverse/notes-from-nature/stats?env=production

File: app/pages/lab/workflow.cjsx [hunk 3]
```diff
@@ -620,7 +620,7 @@ module.exports = React.createClass
       workflowID: ''
 
   render: ->
-    <PromiseRenderer promise={apiClient.type('workflows').get @props.params.workflowID}>{(workflow) =>
+    <PromiseRenderer promise={apiClient.type('workflows').get @props.params.workflowID, {}}>{(workflow) =>
```
Comment (srallen) [inline app/pages/lab/workflow.cjsx]: Is this change intentional?

File: app/pages/project/classify.cjsx [hunk 4]
```diff
@@ -120,7 +121,7 @@ module.exports = React.createClass
         @setState {projectIsComplete}
         randomIndex = Math.floor Math.random() * workflows.length
         # console.log 'Chose random workflow', workflows[randomIndex].id
-        workflows[randomIndex]
+        @getWorkflow project, workflows[randomIndex].id
```
Comment (srallen) [inline app/pages/project/classify.cjsx]: This change is not necessary. All the workflows were already fetched and this just needs to return a random one out of the response array. It doesn't need to be fetched again.

Comment (srallen) [inline app/pages/project/classify.cjsx]: Ah I see the addition of `fields` in the query above, so you'd have to fetch the one you want to get back the whole resource. Ignore my comment.

File: app/lib/get-workflows-in-order.coffee (hunk 1 above)
Comment (srallen) [inline app/lib/get-workflows-in-order.coffee]: Could a comment be left to take the `page_size` query param out of this once pagination is properly implemented?

**Comment by srallen** [comment]
FYI, none of my comments are blockers, so @mcbouslog when you get back, I'll be out of town and if it all looks good to the others, this can go ahead and merge, imo.
